blueocean-git-pipeline/src/main/java/io/jenkins/blueocean/blueocean_git_pipeline/GitReadSaveService.java

```diff
@@ -0,0 +1,259 @@
+                    existingReadCredentialId = existingWriteCredentialId = gitSource.getCredentialsId();
+                    gitTool = Jenkins.getInstance().getDescriptorByType(GitTool.DescriptorImpl.class)
+                            .getInstallation(gitSource.getGitTool());
+                    if (gitTool == null) {
+                        gitTool = GitTool.getDefaultInstallation();
+                    }
+                    for (GitSCMExtension e : gitSource.getExtensions()) {
+                        if (e instanceof GitWritableRepositoryOption) {
+                            GitWritableRepositoryOption writer = (GitWritableRepositoryOption)e;
+                            existingWriteUrl = writer.getRepositoryUrl();
+                            existingWriteCredentialId = writer.getCredentialId();
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+        return new GitReadSaveRequest(
+            StringUtils.defaultIfEmpty(readUrl, existingReadUrl),
+            StringUtils.defaultIfEmpty(writeUrl, StringUtils.defaultIfEmpty(existingWriteUrl, readUrl)),
+            StringUtils.defaultIfEmpty(readCredentialId, existingReadCredentialId),
+            StringUtils.defaultIfEmpty(writeCredentialId, existingWriteCredentialId),
+            StringUtils.defaultIfEmpty(branch, defaultBranch),
+            commitMessage,
+            StringUtils.defaultIfEmpty(sourceBranch, defaultBranch),
+            filePath,
+            contents,
+            gitTool,
+            gitSource
+        );
+    }
+
+    private GitReadSaveRequest makeSaveRequest(Item item, StaplerRequest req) throws IOException, InterruptedException {
+        String readUrl = req.getParameter("readUrl");
+        return makeSaveRequest(
+            item,
+            readUrl,
+            req.getParameter("writeUrl"),
+            req.getParameter("readCredentialId"),
+            req.getParameter("writeCredentialId"),
+            req.getParameter("branch"),
+            req.getParameter("commitMessage"),
+            req.getParameter("sourceBranch"),
+            req.getParameter("path"),
+            Base64.decode(req.getParameter("contents"))
+        );
+    }
+
+    private GitReadSaveRequest makeSaveRequest(Item item, JSONObject json) throws IOException, InterruptedException {
+        JSONObject content = json.getJSONObject("content");
+        return makeSaveRequest(
+            item,
+            null,
+            null,
+            null,
+            null,
+            content.getString("branch"),
+            content.getString("message"),
+            content.getString("sourceBranch"),
+            content.getString("path"),
+            Base64.decode(content.getString("base64Data"))
+        );
+    }
+
+    @Override
+    public Object getContent(StaplerRequest req, Item item) {
+        GitReadSaveRequest r = null;
```

Comment:
make this a `Closable`

so `using(GitReadSaveRequest req = makeSaveRequest(item, req)`